package-res/resources/web/pentaho/type/changes/Changeset.js

```diff
@@ -0,0 +1,187 @@
+     */
+    hasChanges: function() {
+      return this.propertyNames.length > 0;
+    },
+
+    /**
+     * Gets the [change]{@link pentaho.type.changes.ValueChange} object associated with the specified property.
+     *
+     * @param {nonEmptyString|!pentaho.type.Property.Type} name - The property name or type object.
+     *
+     * @return {?pentaho.type.ValueChange} An object describing the changes to be operated
+     * in the given property, or `null` if the property is not defined in this changeset.
+     */
+    getChange: function(name) {
+      var pName = this.owner.type.get(name).name;
+      return pName ? this._properties[pName] || null : null;
+    },
+
+    /**
+     * Sets the value of a property.
+     *
+     * @param {nonEmptyString|!pentaho.type.Property.Type} name - The property name or type object.
+     * @param {(pentaho.type.Value?|pentaho.type.spec.IValue?)} [valueSpec=null] A value specification.
+     */
+    set: function(name, valueSpec) {
+      if(!name) throw error.argRequired("name");
+
+      var owner = this.owner;
+      var pType = owner.type.get(name);
+      var pName = pType.name;
+      var value0 = owner._values[pType.name];
+
+      if(pType.isList) {
+        this._properties[pName] = new ListChange(this.owner, value0, valueSpec);
+      } else {
+        var value1 = pType.toValue(valueSpec);
+
+        if(!pType.type.areEqual(value0, value1)) {
+          this._properties[pName] = new SimpleChange(this.owner, pName, value1);
+        }
+      }
+    },
+
+    /**
+     * Gets the proposed value of a property.
+     *
+     * @param {nonEmptyString|!pentaho.type.Property.Type} name - The property name or type object.
+     *
+     * @return {pentaho.type.Value} The value of the property.
+     */
+    get: function(property) {
+      if(!this.has(property)) return null;
+
+      return this.getChange(property).newValue;
+    },
+
+    /**
+     * Gets the original value of a property.
+     *
+     * @param {nonEmptyString|!pentaho.type.Property.Type} name - The property name or type object.
+     *
+     * @return {pentaho.type.Value} The original value of the property (before the change).
+     */
+    getOld: function(key) {
+      if(!this.has(key)) return null;
+
+      return this.getChange(key).oldValue;
+    },
+
+    /**
+     * Updates the owning [complex]{@linkplain pentaho.type.Complex} with this changes in this changeset.
+     */
+    commit: function() {
+      this._commit();
+    },
+    //endregion
+
+    //region protected methods
+    /**
+     * Prevents further changes to this changeset.
+     * @protected
+     */
+    _freeze: function() {
+      O.eachOwn(this._properties, function(change) {
+        if(change) Object.freeze(change);
+      });
+      Object.freeze(this._properties);
+    },
+
+    /**
+     * Updates the owning [complex]{@linkplain pentaho.type.Complex} with this changes in this changeset.
```

Comment:
suggest:

Updates the owning [complex]{@linkplain pentaho.type.Complex} with the changes in this changeset.